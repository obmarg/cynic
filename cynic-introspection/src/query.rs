pub use queries::*;

#[cynic::schema_for_derives(file = r#"src/schema.graphql"#, module = "schema")]
mod queries {
    use super::schema;

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "Query")]
    pub struct IntrospectionQuery {
        #[cynic(rename = "__schema")]
        pub introspected_schema: IntrospectedSchema,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__Schema")]
    pub struct IntrospectedSchema {
        pub query_type: NamedType,
        pub mutation_type: Option<NamedType>,
        pub subscription_type: Option<NamedType>,
        pub types: Vec<Type>,
        pub directives: Vec<Directive>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__Directive")]
    pub struct Directive {
        pub name: String,
        pub description: Option<String>,
        pub args: Vec<InputValue>,
        pub locations: Vec<DirectiveLocation>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__Type")]
    pub struct Type {
        pub kind: TypeKind,
        pub name: Option<String>,
        pub description: Option<String>,
        #[arguments(includeDeprecated: true)]
        pub fields: Option<Vec<Field>>,
        pub input_fields: Option<Vec<InputValue>>,
        pub interfaces: Option<Vec<NamedType>>,
        #[arguments(includeDeprecated: true)]
        pub enum_values: Option<Vec<EnumValue>>,
        pub possible_types: Option<Vec<NamedType>>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__EnumValue")]
    pub struct EnumValue {
        pub name: String,
        pub description: Option<String>,
        pub is_deprecated: bool,
        pub deprecation_reason: Option<String>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__Field")]
    pub struct Field {
        pub name: String,
        pub description: Option<String>,
        pub args: Vec<InputValue>,
        #[cynic(rename = "type")]
        pub ty: FieldType,
        pub is_deprecated: bool,
        pub deprecation_reason: Option<String>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__InputValue")]

    pub struct InputValue {
        pub name: String,
        pub description: Option<String>,
        #[cynic(rename = "type")]
        pub ty: FieldType,
        pub default_value: Option<String>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__Type")]
    pub struct FieldType {
        pub kind: TypeKind,
        pub name: Option<String>,
        #[cynic(recurse = 5)]
        pub of_type: Option<Box<FieldType>>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "__Type")]
    pub struct NamedType {
        pub name: Option<String>,
    }

    #[derive(cynic::Enum, Clone, Copy, Debug, PartialEq, Eq)]
    #[cynic(graphql_type = "__DirectiveLocation")]
    pub enum DirectiveLocation {
        Query,
        Mutation,
        Subscription,
        Field,
        FragmentDefinition,
        FragmentSpread,
        InlineFragment,
        VariableDefinition,
        Schema,
        Scalar,
        Object,
        FieldDefinition,
        ArgumentDefinition,
        Interface,
        Union,
        Enum,
        EnumValue,
        InputObject,
        InputFieldDefinition,
    }

    #[derive(cynic::Enum, Clone, Copy, Debug)]
    #[cynic(graphql_type = "__TypeKind")]
    pub enum TypeKind {
        Scalar,
        Object,
        Interface,
        Union,
        Enum,
        InputObject,
        List,
        NonNull,
    }
}

mod schema {
    cynic::use_schema!(r#"src/schema.graphql"#);
}
