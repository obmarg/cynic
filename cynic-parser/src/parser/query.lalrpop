use crate::lexer;

use crate::{
    Span,
    executable::{
        storage::*, ids::*, writer::AstWriter
    },
    common::{OperationType, IdRange, WrappingType, TypeWrappers}
};

grammar<'input>(input: &'input str, ast: &mut AstWriter);

pub ExecutableDocument: () = {
    <defs:ExecutableDefinition*> => {}
}

ExecutableDefinition: ExecutableDefinitionId = {
    <def:OperationDefinition> => ast.operation_definition(def),
    <def:FragmentDefinition> => ast.fragment_definition(def),
}

OperationDefinition: OperationDefinitionRecord = {
    <operation_type:OperationType> <name:Name?> <variable_definitions:VariableDefinitions?> <directives:Directives> <selection_set:SelectionSet> => {
        let variable_definitions = ast.variable_definition_range(variable_definitions.map(|defs| defs.len()));
        OperationDefinitionRecord {
            <>
        }
    },
    <selection_set:SelectionSet> => {
        OperationDefinitionRecord {
            operation_type: OperationType::Query,
            name: None,
            selection_set,
            variable_definitions: Default::default(),
            directives: Default::default()
        }
    }
};

FragmentDefinition: FragmentDefinitionRecord = {
    fragment <name:FragmentName> <type_condition:TypeCondition> <directives:Directives> <selection_set:SelectionSet> => {
        FragmentDefinitionRecord {
            <>
        }
    }
};

OperationType: OperationType = {
    query => OperationType::Query,
    mutation => OperationType::Mutation,
    subscription => OperationType::Subscription
}

VariableDefinitions: Vec<()> = {
    "(" <entries:VariableDefinition+> ")" => {
        entries
    }
}

VariableDefinition: () = {
    "$" <name:Name> ":" <ty:Type> <default_value:DefaultValue?> <directives:Directives> => {
        ast.variable_definition(VariableDefinitionRecord { <> });
    }
}

DefaultValue: ValueId = {
    "=" <value:Value> => {
        value
    }
}

SelectionSet: IdRange<SelectionId> = {
    "{" <selections:Selection+> "}" => {
        ast.selection_set(selections)
    }
}

Selection: SelectionRecord = {
    <alias:Alias?> <name:Name> <arguments:Arguments?> <directives:Directives> <selection_set:SelectionSet> => {
        let arguments = ast.argument_range(arguments.map(|args| args.len()));
        SelectionRecord::Field(
            ast.field_selection(FieldSelectionRecord {
                <>
            })
        )
    },
    "..." <name:FragmentName> <directives:Directives> => {
        SelectionRecord::FragmentSpread(
            ast.fragment_spread(FragmentSpreadRecord {
                <>
            })
        )
    },
    "..." <type_condition:TypeCondition?> <directives:Directives> <selection_set:SelectionSet> => {
        SelectionRecord::InlineFragment(
            ast.inline_fragment(InlineFragmentRecord {
                <>
            })
        )
    }
}

Alias: StringId = {
    <alias:Name> ":" => {
        alias
    }
}

TypeCondition: StringId = {
    on <name:NamedType> => {
        name
    }
}

NamedType: StringId = <s:Ident> => ast.ident(s);

Type: TypeId = {
    "[" <ty:Type> => ty,
    <name:NamedType> <wrappers:WrappingType*> => ast.type_reference(TypeRecord {
        name,
        wrappers: TypeWrappers::from_iter(wrappers)
    })
}

WrappingType: WrappingType = {
    "!" => WrappingType::NonNull,
    "]" => WrappingType::List
}


Directives: IdRange<DirectiveId> = {
    <directives:Directive*> => {
        ast.directive_range(Some(directives.len()))
    }
}

Directive: () = {
    "@" <name:Name> <arguments:Arguments?> => {
        let arguments = ast.argument_range(arguments.map(|args| args.len()));
        ast.directive(DirectiveRecord { <> });
    }
}

Arguments: Vec<()> = {
    "(" <arguments:Argument*> ")" => arguments
}

Argument: () = {
    <name:Name> ":" <value:Value> => {
        ast.argument(ArgumentRecord { <> });
    }
}

Value: ValueId = {
    "$" <name:Name> => ast.value(ValueRecord::Variable(name)),
    <int:IntegerLiteral> => ast.value(ValueRecord::Int(int.parse().unwrap())),
    <float:FloatLiteral> => ast.value(ValueRecord::Float(float.parse().unwrap())),
    <s:StringLiteral> => {
        let id = ast.intern_string(s);
        ast.value(ValueRecord::String(id))
    },
    <s:BlockStringLiteral> => {
        let id = ast.intern_string(s);
        ast.value(ValueRecord::String(id))
    },
    true => ast.value(ValueRecord::Boolean(true)),
    false => ast.value(ValueRecord::Boolean(false)),
    null => ast.value(ValueRecord::Null),
    "[" <values:Value*> "]" => ast.value(ValueRecord::List(values)),
    "{" <fields:ObjectField*> "}" => ast.value(ValueRecord::Object(fields)),
    <value:EnumValue> => ast.value(ValueRecord::Enum(value)),
}

ObjectField: (StringId, ValueId) = {
    <name:Name> ":" <value:Value> => (name, value)
}

StringValue: StringId = {
    <s:StringLiteral> => {
        ast.intern_string(s)
    },
    <s:BlockStringLiteral> => {
        ast.intern_string(s)
    },
}

EnumValue: StringId = {
    <s:RawIdent> => ast.ident(s),
    schema => ast.ident("schema"),
    query => ast.ident("query"),
    ty => ast.ident("type"),
    input => ast.ident("input"),
}

Name: StringId = <s:Ident> => ast.ident(s);

FragmentName: StringId = <s:RawFragmentName> => ast.ident(s);

RawFragmentName: &'input str = {
    <s:RawIdent> => s,
    schema => "schema",
    query => "query",
    mutation => "mutation",
    subscription => "subscription",
    ty => "type",
    input => "input",
    true => "true",
    false => "false",
    null => "null",
    implements => "implements",
    interface => "interface",
    "enum" => "enum",
    union => "union",
    scalar => "scalar",
    extend => "extend",
    directive => "directive",
    repeatable => "repeatable",
    fragment => "fragment",
}

Ident: &'input str = {
    <s:RawIdent> => s,
    schema => "schema",
    query => "query",
    mutation => "mutation",
    subscription => "subscription",
    ty => "type",
    input => "input",
    true => "true",
    false => "false",
    null => "null",
    implements => "implements",
    interface => "interface",
    "enum" => "enum",
    union => "union",
    scalar => "scalar",
    extend => "extend",
    directive => "directive",
    repeatable => "repeatable",
    on => "on",
    fragment => "fragment",
}


extern {
    type Location = usize;
    type Error = lexer::LexicalError;

    enum lexer::Token<'input> {
        "$" => lexer::Token::Dollar,

        ":" => lexer::Token::Colon,

        "{" => lexer::Token::OpenBrace,
        "}" => lexer::Token::CloseBrace,

        "(" => lexer::Token::OpenParen,
        ")" => lexer::Token::CloseParen,

        "[" => lexer::Token::OpenBracket,
        "]" => lexer::Token::CloseBracket,

        "!" => lexer::Token::Exclamation,
        "=" => lexer::Token::Equals,
        "@" => lexer::Token::At,
        "&" => lexer::Token::Ampersand,
        "|" => lexer::Token::Pipe,

        "..." => lexer::Token::Spread,

        RawIdent => lexer::Token::Identifier(<&'input str>),

        StringLiteral => lexer::Token::StringLiteral(<&'input str>),
        BlockStringLiteral => lexer::Token::BlockStringLiteral(<&'input str>),
        FloatLiteral => lexer::Token::FloatLiteral(<&'input str>),
        IntegerLiteral => lexer::Token::IntegerLiteral(<&'input str>),

        // Would be nice if these could just be aliases of `Identifier` but LARLPOP doesn't
        // seem to support this well: https://github.com/lalrpop/lalrpop/issues/671
        schema => lexer::Token::Schema,
        query => lexer::Token::Query,
        mutation => lexer::Token::Mutation,
        subscription => lexer::Token::Subscription,
        ty => lexer::Token::Type,
        input => lexer::Token::Input,
        true => lexer::Token::True,
        false => lexer::Token::False,
        null => lexer::Token::Null,
        implements => lexer::Token::Implements,
        interface => lexer::Token::Interface,
        "enum" => lexer::Token::Enum,
        union => lexer::Token::Union,
        scalar => lexer::Token::Scalar,
        extend => lexer::Token::Extend,
        directive => lexer::Token::Directive,
        repeatable => lexer::Token::Repeatable,
        on => lexer::Token::On,
        fragment => lexer::Token::Fragment
    }
}