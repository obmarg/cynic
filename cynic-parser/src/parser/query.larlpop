use crate::lexer;

use crate::ast::{storage::*, ids::*, writer::AstWriter, *};

grammar<'input>(input: &'input str, ast: &mut AstWriter);

pub ExecutableDocument: () = {
    <defs:DefinitionAndDescription+> => {}
}

DefinitionAndDescription: () = {
    <description:StringValue?> <def:ExecutableDefinition> => {
        ast.store_description(def, description)
    }
}

// TODO: Can't re-use DefinitionId here...
ExecutableDefinition: DefinitionId = {
    <def:OperationDefinition> => ast.operation_definition(def),
    <def:FragmentDefinition> => ast.fragment_definition(def),
}

OperationDefinition: OperationDefinition = {
    // TODO
};