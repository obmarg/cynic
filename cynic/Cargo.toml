[package]
name = "cynic"
version = "0.9.0"
authors = ["Graeme Coupar <graeme@polyandglot.dev>"]
edition = "2018"
homepage = "https://cynic-rs.dev"
repository = "https://github.com/obmarg/cynic"
description = "A GraphQL query builder & data mapper for Rust"
keywords = ["graphql", "client", "api"]
license = "MPL-2.0"
documentation = "https://docs.rs/cynic"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
default = []
all = ["chrono", "bson", "uuid", "url"]

[dependencies]
json-decode = "0.5.0"
serde = { version = "1.0.104", features = [ "derive" ] }
serde_json = "1.0"
cynic-proc-macros = { path = "../cynic-proc-macros", version = "0.9.0" }
chrono = { version = "0.4.11", optional = true}
bson = { version = "1.1.0", optional = true}
uuid = { version = "0.8.1", optional = true}
url = { version = "2.1.1", optional = true}

[dev-dependencies]
maplit = "1.0.2"
assert_matches = "1.3.0"
