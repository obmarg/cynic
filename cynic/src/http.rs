//! HTTP client support for cynic.
//!
//! These are hidden behind feature flags by default as HTTP clients are quite
//! heavy dependencies, and there's several options to choose from.

#[cfg(feature = "surf")]
pub use self::surf_ext::SurfExt;

#[cfg(feature = "reqwest")]
pub use reqwest_ext::ReqwestExt;

#[cfg(feature = "reqwest-blocking")]
pub use reqwest_blocking_ext::ReqwestBlockingExt;

#[cfg(feature = "surf")]
mod surf_ext {
    use serde_json::json;
    use std::{future::Future, pin::Pin};

    use crate::{GraphQLResponse, Operation};

    type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;

    /// An extension trait for surf::RequestBuilder.
    ///
    /// ```rust,no_run
    /// # mod query_dsl {
    /// #   cynic::query_dsl!("../schemas/starwars.schema.graphql");
    /// # }
    /// #
    /// # #[derive(cynic::QueryFragment)]
    /// # #[cynic(
    /// #    schema_path = "../schemas/starwars.schema.graphql",
    /// #    query_module = "query_dsl",
    /// #    graphql_type = "Film"
    /// # )]
    /// # struct Film {
    /// #    title: Option<String>,
    /// #    director: Option<String>
    /// # }
    /// #
    /// # #[derive(cynic::QueryFragment)]
    /// # #[cynic(
    /// #     schema_path = "../schemas/starwars.schema.graphql",
    /// #     query_module = "query_dsl",
    /// #     graphql_type = "Root"
    /// # )]
    /// # struct FilmDirectorQuery {
    /// #     #[arguments(id = cynic::Id::new("ZmlsbXM6MQ=="))]
    /// #     film: Option<Film>,
    /// # }
    /// use cynic::{http::SurfExt, QueryBuilder};
    ///
    /// # async move {
    /// let operation = FilmDirectorQuery::build(());
    ///
    /// let response = surf::post("https://swapi-graphql.netlify.com/.netlify/functions/index")
    ///     .run_graphql(operation)
    ///     .await
    ///     .unwrap();
    ///
    /// println!(
    ///     "The director is {}",
    ///     response.data
    ///         .and_then(|d| d.film)
    ///         .and_then(|f| f.director)
    ///         .unwrap()
    /// );
    /// # };
    /// ```
    pub trait SurfExt {
        /// Runs a GraphQL query with the parameters in RequestBuilder, decodes
        /// the and returns the result.
        ///
        /// If a `json_decode::Error` occurs it can be obtained via downcast_ref on
        /// the `surf::Error`.
        fn run_graphql<'a, ResponseData: 'a>(
            self,
            operation: Operation<'a, ResponseData>,
        ) -> BoxFuture<'a, Result<GraphQLResponse<ResponseData>, surf::Error>>;
    }

    impl SurfExt for surf::RequestBuilder {
        fn run_graphql<'a, ResponseData: 'a>(
            self,
            operation: Operation<'a, ResponseData>,
        ) -> BoxFuture<'a, Result<GraphQLResponse<ResponseData>, surf::Error>> {
            Box::pin(async move {
                self.body(json!(&operation))
                    .recv_json::<GraphQLResponse<serde_json::Value>>()
                    .await
                    .and_then(|response| operation.decode_response(response).map_err(|e| e.into()))
            })
        }
    }
}

#[cfg(any(feature = "reqwest", feature = "reqwest-blocking"))]
#[derive(thiserror::Error, Debug)]
pub enum CynicReqwestError {
    #[error("Error making HTTP request: {0}")]
    ReqwestError(#[from] reqwest::Error),
    #[error("Error decoding GraphQL response: {0}")]
    DecodeError(#[from] json_decode::DecodeError),
}

#[cfg(feature = "reqwest")]
mod reqwest_ext {
    use super::CynicReqwestError;
    use std::{future::Future, pin::Pin};

    use crate::{GraphQLResponse, Operation};

    type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;

    /// An extension trait for reqwest::RequestBuilder.
    ///
    /// ```rust,no_run
    /// # mod query_dsl {
    /// #   cynic::query_dsl!("../schemas/starwars.schema.graphql");
    /// # }
    /// #
    /// # #[derive(cynic::QueryFragment)]
    /// # #[cynic(
    /// #    schema_path = "../schemas/starwars.schema.graphql",
    /// #    query_module = "query_dsl",
    /// #    graphql_type = "Film"
    /// # )]
    /// # struct Film {
    /// #    title: Option<String>,
    /// #    director: Option<String>
    /// # }
    /// #
    /// # #[derive(cynic::QueryFragment)]
    /// # #[cynic(
    /// #     schema_path = "../schemas/starwars.schema.graphql",
    /// #     query_module = "query_dsl",
    /// #     graphql_type = "Root"
    /// # )]
    /// # struct FilmDirectorQuery {
    /// #     #[arguments(id = cynic::Id::new("ZmlsbXM6MQ=="))]
    /// #     film: Option<Film>,
    /// # }
    /// use cynic::{http::ReqwestExt, QueryBuilder};
    ///
    /// # async move {
    /// let operation = FilmDirectorQuery::build(());
    ///
    /// let client = reqwest::Client::new();
    /// let response = client.post("https://swapi-graphql.netlify.com/.netlify/functions/index")
    ///     .run_graphql(operation)
    ///     .await
    ///     .unwrap();
    ///
    /// println!(
    ///     "The director is {}",
    ///     response.data
    ///         .and_then(|d| d.film)
    ///         .and_then(|f| f.director)
    ///         .unwrap()
    /// );
    /// # };
    /// ```
    pub trait ReqwestExt {
        /// Runs a GraphQL query with the parameters in RequestBuilder, decodes
        /// the and returns the result.
        ///
        /// If a `json_decode::Error` occurs it can be obtained via downcast_ref on
        /// the `surf::Error`.
        fn run_graphql<'a, ResponseData: 'a>(
            self,
            operation: Operation<'a, ResponseData>,
        ) -> BoxFuture<'a, Result<GraphQLResponse<ResponseData>, CynicReqwestError>>;
    }

    impl ReqwestExt for reqwest::RequestBuilder {
        fn run_graphql<'a, ResponseData: 'a>(
            self,
            operation: Operation<'a, ResponseData>,
        ) -> BoxFuture<'a, Result<GraphQLResponse<ResponseData>, CynicReqwestError>> {
            Box::pin(async move {
                match self
                    .json(&operation)
                    .send()
                    //.recv_json::<GraphQLResponse<serde_json::Value>>()
                    .await
                {
                    Ok(response) => response
                        .json::<GraphQLResponse<serde_json::Value>>()
                        .await
                        .map_err(CynicReqwestError::ReqwestError)
                        .and_then(|gql_response| {
                            operation
                                .decode_response(gql_response)
                                .map_err(CynicReqwestError::DecodeError)
                        }),
                    Err(e) => Err(CynicReqwestError::ReqwestError(e)),
                }
            })
        }
    }
}

#[cfg(feature = "reqwest-blocking")]
mod reqwest_blocking_ext {
    use super::CynicReqwestError;

    use crate::{GraphQLResponse, Operation};

    /// An extension trait for reqwest::blocking::RequestBuilder.
    ///
    /// ```rust,no_run
    /// # mod query_dsl {
    /// #   cynic::query_dsl!("../schemas/starwars.schema.graphql");
    /// # }
    /// #
    /// # #[derive(cynic::QueryFragment)]
    /// # #[cynic(
    /// #    schema_path = "../schemas/starwars.schema.graphql",
    /// #    query_module = "query_dsl",
    /// #    graphql_type = "Film"
    /// # )]
    /// # struct Film {
    /// #    title: Option<String>,
    /// #    director: Option<String>
    /// # }
    /// #
    /// # #[derive(cynic::QueryFragment)]
    /// # #[cynic(
    /// #     schema_path = "../schemas/starwars.schema.graphql",
    /// #     query_module = "query_dsl",
    /// #     graphql_type = "Root"
    /// # )]
    /// # struct FilmDirectorQuery {
    /// #     #[arguments(id = cynic::Id::new("ZmlsbXM6MQ=="))]
    /// #     film: Option<Film>,
    /// # }
    /// use cynic::{http::ReqwestBlockingExt, QueryBuilder};
    ///
    /// let operation = FilmDirectorQuery::build(());
    ///
    /// let client = reqwest::blocking::Client::new();
    /// let response = client.post("https://swapi-graphql.netlify.com/.netlify/functions/index")
    ///     .run_graphql(operation)
    ///     .unwrap();
    ///
    /// println!(
    ///     "The director is {}",
    ///     response.data
    ///         .and_then(|d| d.film)
    ///         .and_then(|f| f.director)
    ///         .unwrap()
    /// );
    /// ```
    pub trait ReqwestBlockingExt {
        /// Runs a GraphQL query with the parameters in RequestBuilder, decodes
        /// the and returns the result.
        ///
        /// If a `json_decode::Error` occurs it can be obtained via downcast_ref on
        /// the `surf::Error`.
        fn run_graphql<'a, ResponseData: 'a>(
            self,
            operation: Operation<'a, ResponseData>,
        ) -> Result<GraphQLResponse<ResponseData>, CynicReqwestError>;
    }

    impl ReqwestBlockingExt for reqwest::blocking::RequestBuilder {
        fn run_graphql<'a, ResponseData: 'a>(
            self,
            operation: Operation<'a, ResponseData>,
        ) -> Result<GraphQLResponse<ResponseData>, CynicReqwestError> {
            self.json(&operation)
                .send()
                .and_then(|response| response.json::<GraphQLResponse<serde_json::Value>>())
                .map_err(CynicReqwestError::ReqwestError)
                .and_then(|gql_response| {
                    operation
                        .decode_response(gql_response)
                        .map_err(CynicReqwestError::DecodeError)
                })
        }
    }
}
