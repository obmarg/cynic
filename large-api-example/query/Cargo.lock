# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "cynic-query"
version = "0.1.0"
