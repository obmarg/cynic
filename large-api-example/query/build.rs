/// Register github schema for creating structs for queries
fn main() {
    cynic_codegen::register_schema("github")
        .from_sdl_file("../../schemas/github.graphql")
        .expect("Failed to find GraphQL Schema");
}