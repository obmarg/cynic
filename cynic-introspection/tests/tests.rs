use cynic_introspection::query::IntrospectionQuery;

#[test]
fn test_introspection_query() {
    insta::assert_snapshot!(build_query().query);
}

fn build_query() -> cynic::Operation<IntrospectionQuery, ()> {
    use cynic::QueryBuilder;

    IntrospectionQuery::build(())
}

#[test]
fn test_running_query() {
    use cynic::http::ReqwestBlockingExt;

    let query = build_query();

    let result = reqwest::blocking::Client::new()
        .post("https://swapi-graphql.netlify.app/.netlify/functions/index")
        .run_graphql(query)
        .unwrap();

    if result.errors.is_some() {
        assert_eq!(result.errors.unwrap().len(), 0);
    }
    insta::assert_debug_snapshot!(result.data);
}

#[test]
fn test_schema_conversion() {
    use cynic::http::ReqwestBlockingExt;

    let query = build_query();

    let result = reqwest::blocking::Client::new()
        .post("https://swapi-graphql.netlify.app/.netlify/functions/index")
        .run_graphql(query)
        .unwrap();

    if result.errors.is_some() {
        assert_eq!(result.errors.unwrap().len(), 0);
    }
    insta::assert_debug_snapshot!(result.data.unwrap().into_schema().unwrap());
}
