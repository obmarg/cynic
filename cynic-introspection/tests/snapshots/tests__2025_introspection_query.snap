---
source: cynic-introspection/tests/tests.rs
expression: build_2025_query().query
snapshot_kind: text
---
query IntrospectionQuery {
  __schema {
    queryType {
      name
    }
    mutationType {
      name
    }
    subscriptionType {
      name
    }
    types {
      kind
      name
      description
      fields(includeDeprecated: true) {
        name
        description
        args(includeDeprecated: true) {
          name
          description
          type {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                  ofType {
                    kind
                    name
                    ofType {
                      kind
                      name
                      ofType {
                        kind
                        name
                      }
                    }
                  }
                }
              }
            }
          }
          defaultValue
          isDeprecated
          deprecationReason
        }
        type {
          kind
          name
          ofType {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                  ofType {
                    kind
                    name
                    ofType {
                      kind
                      name
                    }
                  }
                }
              }
            }
          }
        }
        isDeprecated
        deprecationReason
      }
      inputFields {
        name
        description
        type {
          kind
          name
          ofType {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                  ofType {
                    kind
                    name
                    ofType {
                      kind
                      name
                    }
                  }
                }
              }
            }
          }
        }
        defaultValue
        isDeprecated
        deprecationReason
      }
      interfaces {
        name
      }
      enumValues(includeDeprecated: true) {
        name
        description
        isDeprecated
        deprecationReason
      }
      possibleTypes {
        name
      }
      specifiedByURL
      isOneOf
    }
    directives {
      name
      description
      args(includeDeprecated: true) {
        name
        description
        type {
          kind
          name
          ofType {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                  ofType {
                    kind
                    name
                    ofType {
                      kind
                      name
                    }
                  }
                }
              }
            }
          }
        }
        defaultValue
        isDeprecated
        deprecationReason
      }
      locations
      isRepeatable
    }
  }
}
